src/events/nats.js

```diff
@@ -0,0 +1,170 @@
+        if (jsonConnectProperty) {
+            try {
+                msgJsonStringify = JSON.stringify(msg_internal);
+            } catch (e) {
+                msgJsonStringify = NATS_TYPES.badMessage;
+            }
+        }
+    }
+    return {
+        subject_internal, msg_internal, msgJsonStringify, opt_reply_internal, opt_callback_internal,
+    };
+};
+
+/**
+ * Wrap nats publish function.
+ * @param {Function} original nats publish function.
+ * @param {string} serverHostname nats server host name.
+ * @param {Boolean} jsonConnectProperty json connect property.
+ * @returns {Function} The wrapped function
+ */
+function wrapNatsPublishFunction(original, serverHostname, jsonConnectProperty) {
+    return function internalNatsPublishFunction(subject, msg, opt_reply, opt_callback) {
+        const {
+            subject_internal,
+            msg_internal,
+            msgJsonStringify,
+            opt_reply_internal,
+            opt_callback_internal,
+        } = getPublishParams(subject, msg, opt_reply, opt_callback, jsonConnectProperty);
+        let clientRequest;
+        try {
+            // in case of publish call is a part of request call.
+            if (opt_reply_internal) {
+                return original.apply(this, [subject, msg, opt_reply, opt_callback]);
+            }
+            const { slsEvent: natsPublishEvent, startTime } = eventInterface.initializeEvent(
+                NATS_TYPES.name,
+                serverHostname,
+                'publish',
+                NATS_TYPES.name
+            );
+            const responseMetadata = {
+                subject: subject_internal,
+            };
+            if (!jsonConnectProperty) {
+                responseMetadata.msg = msg_internal;
+            } else if (msgJsonStringify && msgJsonStringify !== NATS_TYPES.badMessage) {
+                responseMetadata.msg = msgJsonStringify;
+            }
+            let patchedCallback = opt_callback_internal;
+
+            if (opt_callback_internal) {
+                const promise = new Promise((resolve) => {
+                    patchedCallback = () => {
+                        eventInterface.finalizeEvent(
+                            natsPublishEvent,
+                            startTime,
+                            null,
+                            responseMetadata
+                        );
+                        resolve();
+                        opt_callback_internal();
+                    };
+                });
+                tracer.addEvent(natsPublishEvent, promise);
+            } else {
+                tracer.addEvent(natsPublishEvent);
+                eventInterface.finalizeEvent(
+                    natsPublishEvent,
+                    startTime,
+                    null,
+                    responseMetadata
+                );
+            }
+            return original.apply(this,
+                [subject_internal, msg_internal, opt_reply_internal, patchedCallback]);
```

Comment:
this seems risky. why not use the original params?

Reply:
just remove this call to outside of the try / catch. currently, you are never assigning clientRequest and if the original function throws an error you call it twice!!

Reply:
Because I already build the right params. But maybe you right and its better to pass the original params, in case of nats change is api.

Reply:
instead of return this original.apply, I assign it as ClientRequest, and changed the three first params to the original params.